### PR TITLE
Trigger API change

### LINKED ISSUE #230: Structured trigger format
## Overview

The design for #217 comes with two issues:
- Validation can get complicated as the number of triggers increases
- The CRD cannot express what fields are available for each trigger type

## Design

To solve these problems, we should use a technique similar to what [Argo Event TriggerTemplate](https://github.com/argoproj/argo-events/blob/master/api/sensor.md#triggertemplate) does.

```yaml
apiVersion: pipelines.kubeflow.org/v1alpha5
kind: RunConfiguration
metadata:
  name: ...
spec:
  run:
    ...
  triggers:
  - schedule:
      cronExpression: "1 2 3 4 5"
  - change: {}
```

Validation should be changed to allow only one trigger "type" (field).

### PR DESCRIPTION
**Pull request by jmendesky**
Closes #230 

## Tasks

- [x] Documentation updated

### CODE REVIEW

File: apis/pipelines/v1alpha5/runconfiguration_types.go [hunk 1]
```diff
@@ -7,29 +7,39 @@ import (
...
-	Type TriggerType `json:"type"`
-	// +kubebuilder:validation:Required
-	CronExpression string `json:"cronExpression"`
+	Schedule *ScheduleTrigger `json:"schedule,omitempty"`
+	OnChange *OnChangeTrigger `json:"onChange,omitempty"`
 }
 
-type TriggerType string
+type ScheduleTrigger struct {
+	CronExpression string `json:"cronExpression"`
```
Comment (paoloambrosio-skyuk) [inline apis/pipelines/v1alpha5/runconfiguration_types.go:15]: Not too important though.
```suggestion
	// +kubebuilder:validation:Required
	CronExpression string `json:"cronExpression"`
```

File: apis/pipelines/v1alpha5/test_generators.go [hunk 2]
```diff
@@ -48,12 +49,37 @@ func RandomRunConfiguration() *RunConfiguration {
...
+		Triggers: RandomList(RandomTrigger),
+	}
+}
+
+func RandomTrigger() Trigger {
+	switch common.RandomInt64() % 2 {
+	case 0:
+		return RandomCronTrigger()
+	case 1:
+		return Trigger{OnChange: &OnChangeTrigger{}}
```
Comment (paoloambrosio-skyuk) [inline apis/pipelines/v1alpha5/test_generators.go:61]:
```suggestion
		return RandomOnChangeTrigger()
```

**Comment by jmendesky** [comment]
ok to test

File: apis/pipelines/v1alpha5/runconfiguration_webhook_test.go [hunk 3]
```diff
@@ -21,42 +21,54 @@ var _ = Context("RunConfiguration Webhook", func() {
...
 			errors := rc.errorsInTriggers()
 			Expect(errors).To(HaveLen(1))
 			Expect(errors[0].Type).To(Equal(field.ErrorTypeRequired))
-			Expect(errors[0].Field).To(Equal("spec.triggers[0].cronExpression"))
+			Expect(errors[0].Field).To(Equal("spec.triggers[0].schedule.cronExpression"))
 			Expect(errors[0].Detail).To(Equal("required for trigger type schedule"))
 		})
 
-		Specify("rejects cron expression for non-schedule triggers", func() {
+		Specify("rejects more than one trigger", func() {
```
Comment (jmendesky) [inline apis/pipelines/v1alpha5/runconfiguration_webhook_test.go:48]:
```suggestion
		Specify("rejects more than one trigger type in the same trigger", func() {
```

File: apis/pipelines/v1alpha5/runconfiguration_webhook_test.go [hunk 4]
```diff
@@ -21,42 +21,54 @@ var _ = Context("RunConfiguration Webhook", func() {
...
+			}
+
+			errors := rc.errorsInTriggers()
+			Expect(errors).To(HaveLen(1))
+			Expect(errors[0].Type).To(Equal(field.ErrorTypeTooMany))
+			Expect(errors[0].Field).To(Equal("spec.triggers[0]"))
+			Expect(errors[0].Detail).To(Equal("must have at most 1 items"))
+		})
+
+		Specify("rejects no trigger", func() {
```
Comment (jmendesky) [inline apis/pipelines/v1alpha5/runconfiguration_webhook_test.go:61]:
```suggestion
		Specify("rejects no trigger type in a trigger", func() {
```

File: apis/pipelines/v1alpha5/runconfiguration_webhook_test.go [hunk 5]
```diff
@@ -21,42 +21,54 @@ var _ = Context("RunConfiguration Webhook", func() {
...
 			}
 
 			errors := rc.errorsInTriggers()
 			Expect(errors).To(HaveLen(1))
-			Expect(errors[0].Type).To(Equal(field.ErrorTypeForbidden))
-			Expect(errors[0].Field).To(Equal("spec.triggers[1].cronExpression"))
-			Expect(errors[0].Detail).To(Equal("not allowed for trigger type onChange"))
+			Expect(errors[0].Type).To(Equal(field.ErrorTypeRequired))
+			Expect(errors[0].Field).To(Equal("spec.triggers[0]"))
+			Expect(errors[0].Detail).To(Equal("a trigger must be set"))
```
Comment (jmendesky) [inline apis/pipelines/v1alpha5/runconfiguration_webhook_test.go:71]: as above: 
```suggestion
			Expect(errors[0].Detail).To(Equal("a trigger type must be set"))
```

**Review by paoloambrosio-skyuk — APPROVED** [review]

**Pull request merged**